alliance/share/man/man1/genpat.1
.\"   $Id: genpat.1,v 1.2 1999/09/24 14:12:13 simcity Exp $
.\" @(#)genpat 2.0 Sep 07 1993 UPMC; Author: PITON D.
.TH GENPAT 1 "October 1, 1997" "ASIM/LIP6" "ALLIANCE USER COMMANDS"

.SH NAME
.PP
\fBgenpat\fP, A procedural pattern file generator

.SH SYNOPSIS
.PP
\fBgenpat [-v] [-k] [file]\fP
.br

.so man1/alc_origin.1

.SH DESCRIPTION
.PP
\fBGenpat\fP is a set of C fonctions that allows a procedural description of
input pattern file for the logic simulator \fBASIMUT\fP. The Unix \fBgenpat\fP
command accepts a C file as input and produces a pattern description file as
output. The extension ".c" is not to be given. The file generated by
\fBgenpat\fP is in pat format, so IT IS STRONGLY RECOMMENDED TO SEE pat(5)
BEFORE THIS MANUAL.

.SH OPTIONS
.PP
.TP 15
\fI\-v\fP
verbose mode
.TP 15
\fI\-k\fP
keeps the executable along with the compilation Makefile after completion

.SH GENPAT FILE FORMAT
.PP
From a user point of view, \fBgenpat\fP is a pattern description language
using all standard C facilities (include, define, variables, loop, ...).
Fonctions provided by \fBgenpat\fP are to be used in a given order. Using
them in a different order won't crash the system, but will result in
execution errors. Here follows the description of the input file.

.PP
A \fBpat\fP format file can be divided in two parts : declaration and
description part.

.PP
The declaration part is the list of inputs, outputs, internal signals and
registers. Inputs are to be forced to a certain value and all the others are
to be observed during simulation.

.PP
The description part is a series of patterns, where each pattern defines the
value of inputs and outputs at a given date. This date represents the absolute
time for the simulator.

.PP
Similarly, a \fBgenpat\fP file can be divided in two parts : declaration and
description part. Functions related to the declaration must be called before
any function related to the description part.

.PP
.TP 10
\fIdeclaration part\fP
The first thing you should do in this part is to give the output file's name
(see DEF_GENPAT(3)). Then, this part allows you to declare the inputs, the
outputs, and internal observing points (see DECLAR(3)). It is also possible
to create virtual arraies (see ARRAY(3)).
.TP 10
\fIdescription part\fP
After all signals are declared, you can begin to define input values which
are to be applied to the inputs of the circuit at given simulation time.
You can also define output values which are to be compare with the values
produced during the simulation. (see AFFECT(3)).
Genpat describes the stimulus by event : only signal transitions are described.
Genpat manage the simulation time in ps.
This part also allows you to give instructions to the simulation tool to save
the state of the circuit at the end of the simulation. (see SAVE(3)). Last
thing you should do in this part is to generate the output file (see
SAV_GENPAT(3)).  

.SH FUNCTIONS
.PP
.TP 15
\fIDEF_GENPAT()\fP 
defines the output file's name.
.TP 15   
\fISAV_GENPAT()\fP
make the output file be generated
.TP 15
\fIDECLAR()\fP
declares inputs, outputs, and the internal observing points.
.TP 15
\fIARRAY()\fP
allows signals of the same type to be groupped in an "virtual array" in
order to ease their manipulation 
.TP 15
\fIINIT()\fP
changes the values of registers between two patterns.
.TP 15
\fIAFFECT()\fP
assigns a value to a signal, at a given simulation time (in ps). This value
is kept on the signal until a new value is assigned to the signal.
.TP 15
\fISAVE()\fP
informs the simulation tool to save the state of the circuit at the end of
simulation
.TP 15
\fILABEL()\fP
gives a label to the current pattern.
.TP 15
\fIGETCPAT()\fP
return the date of the current pattern (in ps).

.SH EXAMPLES
.PP
 
.nf
#include <stdio.h>
#include "genpat.h"

char *inttostr(entier)
int entier;
  {
  char *str;
  str = (char *) mbkalloc (32 * sizeof (char));
  sprintf (str, "%d",entier);
  return(str);
  }
 
 main () 
 {
 int i;
 int j;
 int vect_date; /* this date is an absolute date, in ps */
 
 DEF_GENPAT("vecteurs");
 
 /* interface */
 DECLAR ("a", ":2", "X", IN, "0  to 3");
 DECLAR ("b", ":2", "X", IN, "0  to 3");
 DECLAR ("sel", ":2", "B", IN, "");
 DECLAR ("ck", ":2", "B", IN, "");
 DECLAR ("s", ":2", "X", OUT, "0 to 3");
 
 LABEL ("additionneur");
 AFFECT ("0", "ck", "0b0");
 AFFECT ("0", "sel", "0b0");
 
 for (i=0; i<16; i++)
     for (j=0; j<16; j++) 
     {
         vect_date = i*16 + j;
         AFFECT (inttostr(vect_date), "a", inttostr(i));
         AFFECT (inttostr(vect_date), "b", inttostr(j));
     }
 
 vect_date = vect_date + 1;
 AFFECT (inttostr(vect_date), "a", "0b0");
 AFFECT (inttostr(vect_date), "b", "0b0");
 AFFECT (inttostr(vect_date), "sel", "0b0");
 AFFECT (inttostr(vect_date), "ck" , "0b0");
 
 SAV_GENPAT ();
 }
.fi
 
.SH ENVIRONMENT VARIABLES
.PP
\fBGenpat\fP reads the environment variable VH_PATSFX to give the result
file an extension.

.SH SEE ALSO
.PP
AFFECT(3), ARRAY(3), DECLAR(3), DEF_GENPAT(3), GETCPAT(3), INIT(3), LABEL(3),
SAVE(3), SAV_GENPAT(3), pat(5), asimut(1)


.so man1/alc_bug_report.1



alliance/CHANGES
# Alliance VLSI CAD System
#
# Home page          : http://www-asim.lip6.fr/alliance/
# E-mail support     : mailto:alliance-support@asim.lip6.fr
# ftp sites          : ftp://ftp.lip6.fr/lip6/softs/alliance/
#                   or ftp://ftp-asim.lip6.fr/pub/alliance/
#
# $Id: CHANGES,v 1.3 2000/01/19 18:35:30 czo Exp $

--------------------------------------------------------------------------------

ALLIANCE release 4.0 (15/10/99)

1/ New Standard Cells Library : sxlib
   sxlib allows multi-layer over cell routing
   and all transitions are now on rising edge (man sxlib)

2/ Improved Tools :
   - asimut is now a temporal logic simulator. 
     You *must* modify your old pattern files to allow delays (man asimut)
   - syf can now handle vbe statements in .fsm
   - dreal : New viewing options ( like -install for private colormap)
   - graal : New viewing options ( like -install for private colormap)
             Input scale changeable, Hierarchy browser, real view (.cif/.gds)
   
3/ New Tools :
   algue : ALliance Graphic User Environement
   ali   : gives information on Alliance Environnement
   b2f   : FSM Abstraction from .vbe
   k2f   : Translate kiss2 format <-> FSM format
   vasy  : Translate .vhd format (Synopsys) <-> .vbe format 
   xsch  : Netlist viewer
   xfsm  : FSM viewer
   xpat  : patterns (with delay) viewer


--------------------------------------------------------------------------------

ALLIANCE release 3.5 (19/06/98)

1/ Lots of improvments and bug fixes

2/ AP : Now supports ten layers of metal

3/ AL : New support for RC net in losig. Parsers al, spice, vti modified.
   Modified Data Structures :
    - lotrs : added field BULK TRNAME
    - locon : added field PNODE
    - losig : added field PRCN, CAPA deleted
    
   Modified functions (the CAPA field has moved) :
        addlosig, addlotrs, dellosig

4/ RCN : New library for RC support : Resistances on losig and capcitances
   beetwen losig.
   
--------------------------------------------------------------------------------

ALLIANCE release 3.2c (20/03/98)

1/ Ap format :

   Alliance ap format now supports 
   - half grid spacing
   - connectors somewhere else than abutment box.

2/ VERILOG driver :

   Support added for Alliance-Cadence Toolbox

3/ MBK_SCALE_X now defaults to 100

--------------------------------------------------------------------------------

ALLIANCE release 3.2b (15/12/97)

1/ Easy install

   A "configure" script is now available to configure
   Alliance on any UNIX system

2/ New driver :

   Support to VERILOG netlist as been added.
   Only driver exits. This means you can save your 
   netlists to VERILOG format 'vlg'

3/ New names :

  - Logic has been splited in 3 parts,
    . bop   : boolean optimizer (logic -o) 
    . scmap : Std cell mapping (logic -s) 
    . c4map : CCCC mapping (logic -c) 

  - Desb is replaced by yagle

  - Alligator is replaced by fpmap (X4000)

  - Netoptim is replaced glop

4/ Cells libraries

   The tree of directories containing the cells 
   libraries has been simplified

--------------------------------------------------------------------------------

ALLIANCE release 3.2 (17/05/97)


1/ GRAPHICAL PATTERN VIEWER

   In order to see the patterns resulting from a simulation,
   the XPAT tools has been developped.

2/ GRAPHICAL FSM VIEWER

   In order to see the state's graph of an FSM the XFSM tools has 
   been developped.

3/ RECTANGLE LAYOUT VIEWER

   Now, DREAL is also a real layout editor.

--------------------------------------------------------------------------------

ALLIANCE release 3.0 (17/05/94)

1/ FPGA SYNTHESIS

   A logic synthesis tools that maps on FPGA is now available.
   It works for X3000 devices

2/ FLOOR-PLAN ROUTING

   Binaries of the CHEOPS router from BULL are available for sparc.
   
3/ TIMING ANALYSIS

   The static timing analisys tools TAS is finally available.
   It can be targetted to several processes though the use of a technological
   file suffixed `elp'.

4/ RECTANGLE LAYOUT VIEWER

   In order to see the layout resulting from a symbolic to real translation,
   the DREAL tools has been developped.
   
--------------------------------------------------------------------------------
ALLIANCE release 2.0 (14/02/94) versus ALLIANCE 1.2

1/ SYMBOLIC LAYOUT EDITOR

  The symbolic layout editor ALC has been replaced by GRAAL.
  GRAAL provides the same functionnalities than ALC, but is much
  more reliable. GRAAL support both CMOS and GaAs symbolic layout.

> man graal

2/ DESIGN RULE CHECKER

  The Design Rule Checker VERSATIL has been replaced by DRUC.
  DRUC provides the same functionnalities than VERSATIL.
  A hierarchical version will be distributed in the next ALLIANCE release.

> man druc

3/ LOGIC SYNTHESIS

  The logic synthesis tool LOGIC has been strongly improved.
  The new tool NETOPTIM is a gate-level net-list optimiser
  that allows to minimize delays in a synthesized gate net-list.
  The Finite-state-machine synthesizer SYF allows to describe and
  synthesize high complexity FSM (more than 100 states)
  It is possible to describe hierachical FSM using stack  (subroutines).

> man logic
> man syf
> man netoptim

4/ DATA-PATH COMPILER

  FPGEN is a data-path compiler using a dedicated macro-cells library.
  DPR is the place and route tool that creates optimized data-path blocks
  from the gate net-list generated by FPGEN.

> man fpgen
> man dpr

5/ PARAMETERIZED MACRO-CELLS

  Six parametrized generators are part of this release:

> man rsa                # fast adder generator
> man bsg                # barrel shifter generator
> man amg                # multiplier generator
> man rfg                # register file generator
> man grog               # high speed ROM generator
> man rage               # static RAM generator

6/ PROCEDURAL LAYOUT DEBUGGER

  The graphic debugger GENVIEW allows to debug custom blocks
  described with the procedural language GENLIB.

alliance/src/genpat/src/genpat.sh
#!/bin/sh 
# $Id: genpat.sh,v 1.2 2002/10/24 10:30:25 hcl Exp $

#set -v 
#set -x

help() {
      echo "Syntax: `basename $0` [-vk] source-file (without extension)"
      echo "                       -v : verbose mode"
      echo "                       -k : keeps  the executable (whith debugging"
      echo "                            informations) along with the"
      echo "                            compilation Makefile after completion"
      exit 1
}

   if [ $# -lt 1 -o $# -gt 4 ] ; then
    help 1 2 
   fi

   debug=
   talk=0
   keep=0
   name=""
	while [ $# -gt 0 ]
	do
		case $1 in
      -v)  talk=1;;
      -k)  keep=1;debug=-g;;
      -vk) keep=1;debug=-g; talk=1;;
      -kv) keep=1;debug=-g; talk=1;;
      *)  if [ -z "$name" ]; then 
             name=$1;
          else
	    help
	  fi
      esac
      shift
	done
 	trap "rm -f ./*.$$; exit 0" 1 2 3 6

	alcbanner "GenPat" "3.1" "Procedural GENeration of test PATterns" "1991"

	if [ -z "$name" ] ; then
	    help
	fi
	if [ ! -f $name.c ] ; then
		echo "There seems no to be a file called $name.c"
		help
	fi
		
	if [ $talk -eq 1 ]; then
		echo "Generating the Makefile";
	fi

	makefile="Makefile.$$"


cat << EOF  > $makefile

ALLIANCE_LIBRARY = -L\$(ALLIANCE_TOP)/lib \\
                     -lPgn \\
                     -lPpt \\
                     -lPhl \\
                     -lPat \\
                     -lMut \\
                     -lRcn


ALLIANCE_INC = -I\$(ALLIANCE_TOP)/include 

$name : $name.c
	\$(CC) $debug -o $name $name.c \$(ALLIANCE_INC) \$(ALLIANCE_LIBRARY)

EOF

   if [ $talk -eq 1 ]; then 
		echo "Compiling, ..."
	fi

   make -f $makefile > $name.grr 2>&1

   if [ ! $? -eq 0 ]; then 
      echo "Compilation failed!"
      cat $name.grr
		exit 1
	fi

   if [ $talk -eq 1 ] ; then
	   echo "Current execution environment"
		echo "MBK_CATA_LIB   : ${MBK_CATA_LIB-no cell library specified}"
		echo "MBK_WORK_LIB   : ${MBK_WORK_LIB-:}"
		echo "MBK_IN_LO      : ${MBK_IN_LO-vst}"
		echo "MBK_OUT_LO     : ${MBK_OUT_LO-vst}"
		echo "MBK_IN_PH      : ${MBK_IN_PH-ap}"
		echo "MBK_OUT_PH     : ${MBK_OUT_PH-ap}"
		echo "MBK_CATAL_NAME : ${MBK_CATAL_NAME-CATAL}"
   fi

   if [ $talk -eq 1 ]; then
      echo "Executing ..."
	fi
   ./$name
   exit_code=$?

   if [ $talk -eq 1 ]; then
		echo "Removing tmp files ..."
	fi

   if [ $keep -eq 0 ] ; then
		rm $name $name.exe;
	fi

   rm $name.o $name.grr $makefile > /dev/null 2>&1

   exit $exit_code

